### PR TITLE
Ingest XDP v0.16.7 and use TCP CIBIR rules when needed

### PR DESCRIPTION
**Pull request by csujedihy**
## Description

Ingest XDP v0.16.7 and use TCP CIBIR rules when needed.

Changes to our tests:
1. 0-RTT tests are disabled. QTIP caches only one packet before handshake completes. So, if multiple 0-RTT packets are sent, only one of them will make it.
2. NAT rebind doesn't apply to TCP.  So, they are disabled.
3. Due to implementation limitations, sharing bindings isn't supported. Some tests are tweaked to always create a new binding per connection. For simplicity, the implementation assumes client side sockets have a 1:1 mapping to QUIC connection.

## Testing

CI

### CODE REVIEW

**Comment by csujedihy** [comment]
@nibanks do you know what secnetperf commands I can use to test this end to end?

**Comment by nibanks** [comment]
> @nibanks do you know what secnetperf commands I can use to test this end to end? Is it just the client and server that use the same hex passed in `-cibir:` can talk to each other?

Yeah, that should work.

**Comment by csujedihy** [comment]
Oh damn, it doesn't really simply work this way. The handshake won't be completed because the handshake packets don't have "UDP" payload and won't be picked up by CIBIR rules.

**Comment by nibanks** [comment]
> I am sorry I didn't foresee this. :/

No problem. 😃 None of us did.

I'll have to think about the options here. I'd prefer less changes on the XDP side, and more on the QUIC side to work around, so it doesn't possibly effect non-QTIP stuff on the XDP side.

**Comment by csujedihy** [comment]
> Oh damn, it doesn't really simply work this way. The handshake won't be completed because the handshake packets don't have "UDP" payload and won't be picked up by CIBIR rules.



> > I am sorry I didn't foresee this. :/
> 
> No problem. 😃 None of us did.
> 
> I'll have to think about the options here. I'd prefer less changes on the XDP side, and more on the QUIC side to work around, so it doesn't possibly effect non-QTIP stuff on the XDP side.

Sync'd up with Michael. I am going with option 1.

**Comment by nibanks** [comment]
I've tried adding the test code, but something really wonky is going on with `QUIC_API_ENABLE_PREVIEW_FEATURES` definition on Windows. I don't understand why it's not getting set... Any ideas?

**Comment by csujedihy** [comment]
> > Oh damn, it doesn't really simply work this way. The handshake won't be completed because the handshake packets don't have "UDP" payload and won't be picked up by CIBIR rules.
> 
> > > I am sorry I didn't foresee this. :/
> > 
> > 
> > No problem. 😃 None of us did.
> > I'll have to think about the options here. I'd prefer less changes on the XDP side, and more on the QUIC side to work around, so it doesn't possibly effect non-QTIP stuff on the XDP side.
> 
> Sync'd up with Michael. I am going with option 1.

Pushed a new iteration with the new XDP rule. CIBIR is working now.

File: scripts/build.ps1 [hunk 1]
```diff
@@ -283,6 +283,7 @@ if (!$OfficialRelease) {
         # Thanks to https://stackoverflow.com/questions/3404936/show-which-git-tag-you-are-on
         # for this magic git command!
         $Output = git describe --exact-match --tags $(git log -n1 --pretty='%h')
+        Write-Host $Output
```
Comment (nibanks) [inline scripts/build.ps1:286]: Remove this now?

File: src/test/lib/DataTest.cpp [hunk 2]
```diff
@@ -14,6 +14,10 @@
 #include "DataTest.cpp.clog.h"
 #endif
 
+#ifdef QUIC_USE_RAW_DATAPATH
+extern bool UseQTIP;
+#endif
```
Comment (nibanks) [inline src/test/lib/DataTest.cpp:19]: Would it be better to make this bool always present but only ever set to true when RAW is used? That'd simplify all these #ifdefs

File: src/test/lib/MtuTest.cpp [hunk 3]
```diff
@@ -45,6 +49,12 @@ struct ResetSettings {
 void
 QuicTestMtuSettings()
 {
+#ifdef QUIC_USE_RAW_DATAPATH
+    uint16_t DefaultMaximumMtu = UseQTIP ? 1488 : 1500; // reserve 12B for TCP header
+#else
+    uint16_t DefaultMaximumMtu = 1500;
+#endif
```
Comment (nibanks) [inline src/test/lib/MtuTest.cpp:56]: Make them `const`?

File: src/test/lib/MtuTest.cpp [hunk 4]
```diff
@@ -45,6 +49,12 @@ struct ResetSettings {
 void
 QuicTestMtuSettings()
```
Comment (nibanks) [inline src/test/lib/MtuTest.cpp:50]: This test is really fragile. It assumes a max MTU that we might change in the future...

File: src/test/lib/TestHelpers.h [hunk 5]
```diff
@@ -46,6 +46,10 @@ QuicAddrSetToDuoNic(
     }
 }
 
+#ifndef QUIC_OFFICIAL_RELEASE
+#define QUIC_API_ENABLE_PREVIEW_FEATURES
+#endif
```
Comment (nibanks) [inline src/test/lib/TestHelpers.h:51]: We should be able to remove this now. I added it when I was trying to figure out what was going on.

Comment (csujedihy) [reply src/test/lib/TestHelpers.h]: removed

File: src/test/lib/MtuTest.cpp (hunk 3 above)
Comment (csujedihy) [reply src/test/lib/MtuTest.cpp]: done

File: src/test/lib/DataTest.cpp (hunk 2 above)
Comment (csujedihy) [reply src/test/lib/DataTest.cpp:19]: That can't be shared across user and kernel mode.

File: .azure/azure-pipelines.ci.yml [hunk 6]
```diff
@@ -659,6 +659,17 @@ stages:
       platform: windows
       tls: schannel
       logProfile: Full.Light
+  - template: ./templates/run-bvt.yml
```
Comment (csujedihy) [inline .azure/azure-pipelines.ci.yml:662]: How can I get this to work?

Comment (csujedihy) [reply .azure/azure-pipelines.ci.yml:662]: I am getting this error: `Stage test_bvt job bvt_windows_x64_schannel_false_false_windows_2022_Xdp appears more than once. Job names must be unique within a stage.`

Comment (nibanks) [reply .azure/azure-pipelines.ci.yml:662]: I think there's an arg like extra name you need to set to something unique

Comment (csujedihy) [reply .azure/azure-pipelines.ci.yml:662]: Ah, I see. The parameters form the job name in run-bvt.yml. However, the job name doesn't include extraName. I will remove this for name and handle it in a separate PR after getting the github actions stable.

**Comment by csujedihy** [comment]
The only failing test cases are NAT rebind cases. These don't make sense for TCP.

**Review by nibanks — APPROVED** [review]

**Review by nibanks — APPROVED** [review]

**Pull request merged**